### DISCUSSION TITLE
Redirect console output to string

### DISCUSSION THREAD

**Discussion by SashmanGH** [coin-or/Sonnet#42] [Q&A] [answered]
Hi,
I would like to capture all the solver out that is written to the console in a string. I tried to use Console.SetOut to set it to a StringWriter just before calling the Minimise-method, but the ToString-call after the Minimise returned an empty string. 

Is there anything I can do to either caputre the complete console output or to redirect it to a file so I can read it afterwards?

Thanks,
Sascha

**Comment by jhmgoossens** [comment, marked as answer]
Hi,
Sonnet is a .NET (Framework) library, but the actual solving is done in the coin-or libraries like Cbc etc. in SonnetWrapper. 
Sonnet itself outputs to the regular System.Console of .NET.
However, the underlying coin-or C++ libraries like Cbc use (s)printf etc. via message handlers to output, e.g. in [CbcSolver.cpp](https://github.com/coin-or/Cbc/blob/209c1e8eb143ea67360aa4578a44e9d0373718b1/src/CbcSolver.cpp#L1548). This output from (s)printf is not going to System.Console by itself.

Sonnet tries to hand a wrapped MessageHandler to the underlying coin-or C++ libraries, for example when [reading Mps files](https://github.com/coin-or/Sonnet/blob/0d6fb2c20e07e6794b4337ea2792653760ed3f50/src/Sonnet/Model.cs#L330) to get the underlying message over to System.Console, but this isn't working for much of the CbcSolver output, which is why your Console.SetOut isn't capturing it, unfortunately. I haven't been able to figure out how to pass more message to the Sonnet MessageHandler--you can use `SonnetLog.Default.PassToSolver(solver.OsiSolver)` before calling `solver.Minimise()`, which grabs a bit more output.

A completely different way to combine all output is to use the Process class and [Process.StandardOutput](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.standardoutput?view=netframework-4.8.1), but this requires that your Sonnet code is in a separate executable from your main application. 

Lastly, if what you're trying to do is keep progress, then you could rather use some of the Cbc (or Clp etc.) EventHandlers that are called every time a new solution is found, etc. See [Sonnet_CbcTests](https://github.com/coin-or/Sonnet/blob/0d6fb2c20e07e6794b4337ea2792653760ed3f50/test/SonnetTest/Sonnet_CbcTests.cs#L172) for examples of how to inject handlers from Sonnet. Or, similarly, the CbcSolver CallBack.

Hope this help?

**Reply by SashmanGH** [reply]
Hi,
great, that helps. I got the EventHandler working (as decribed in the cbc tests). The separate executable approach is not suitable for my application. I tried the SonnetLog, but I couldn't get it to work. I called PassToSolver, but then I didn't know how to get the logged messages from the SonnetLog. If I read the sources correctly I believe I can only write to the log, but not read it. Or am I on the wrong track?

Thanks,
Sascha

**Comment by jhmgoossens** [comment]
Great!

> but then I didn't know how to get the logged messages from the SonnetLog

SonnetLog writes to System.Console, so SonnetLog output you can read using Console.SetOut. Does that work for you? (But, again, not all cbc message are nicely passed to the SonnetLog message handler..)

**Reply by SashmanGH** [reply]
Hi,
that did the trick. Many thanks!
Sascha
